matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateRoomFromLocalRoomTask.kt

```diff
@@ -0,0 +1,303 @@
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+
+/**
+ * Create a Room from a "fake" local room.
+ * The configuration of the local room will be use to configure the new room.
+ * The potential local room members will also be invited to this new room.
+ *
+ * A "fake" local tombstone event will be created to indicate that the local room has been replacing by the new one.
+ */
+internal interface CreateRoomFromLocalRoomTask : Task<CreateRoomFromLocalRoomTask.Params, String> {
+    data class Params(val localRoomId: String)
+}
+
+internal class DefaultCreateRoomFromLocalRoomTask @Inject constructor(
+        @UserId private val userId: String,
+        @SessionDatabase private val monarchy: Monarchy,
+        private val createRoomTask: CreateRoomTask,
+        private val stateEventDataSource: StateEventDataSource,
+        private val clock: Clock,
+) : CreateRoomFromLocalRoomTask {
+
+    private val realmConfiguration
+        get() = monarchy.realmConfiguration
+
+    override suspend fun execute(params: CreateRoomFromLocalRoomTask.Params): String {
+        val replacementRoomId = stateEventDataSource.getStateEvent(params.localRoomId, EventType.STATE_ROOM_TOMBSTONE, QueryStringValue.IsEmpty)
+                ?.content?.toModel<RoomTombstoneContent>()
+                ?.replacementRoomId
+
+        if (replacementRoomId != null) {
+            return replacementRoomId
+        }
+
+        val createRoomParams = getCreateRoomParams(params)
+        val roomId = createRoomTask.execute(createRoomParams)
+
+        try {
+            // Wait for all the room events before triggering the replacement room
+            awaitNotEmptyResult(realmConfiguration, TimeUnit.MINUTES.toMillis(1L)) { realm ->
+                realm.where(RoomSummaryEntity::class.java)
+                        .equalTo(RoomSummaryEntityFields.ROOM_ID, roomId)
+                        .equalTo(RoomSummaryEntityFields.INVITED_MEMBERS_COUNT, createRoomParams.invitedUserIds.size.minus(1))
+            }
+            awaitNotEmptyResult(realmConfiguration, TimeUnit.MINUTES.toMillis(1L)) { realm ->
+                EventEntity.whereRoomId(realm, roomId)
+                        .equalTo(EventEntityFields.TYPE, EventType.STATE_ROOM_HISTORY_VISIBILITY)
+            }
+            if (createRoomParams.algorithm != null) {
+                awaitNotEmptyResult(realmConfiguration, TimeUnit.MINUTES.toMillis(1L)) { realm ->
+                    EventEntity.whereRoomId(realm, roomId)
+                            .equalTo(EventEntityFields.TYPE, EventType.STATE_ROOM_ENCRYPTION)
+                }
+            }
+        } catch (exception: TimeoutCancellationException) {
+            throw CreateRoomFailure.CreatedWithTimeout(roomId)
+        }
+
+        createTombstoneEvent(params, roomId)
+        return roomId
+    }
+
+    /* ==========================================================================================
```

Comment:
I think we agreed to not use section bloc comments. They could be removed if we split this class into several components.